File: sort/56.merge-intervals.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) < 2:
            return intervals

        intervals.sort(key=lambda x: x[0])
        ans = []
        for interval in intervals:
            l, r = interval[0], interval[1]
            if not ans or ans[-1][1] < l:
                ans.append([l, r])
            else:
                ans[-1][1] = max(ans[-1][1], r)

        return ans


print(Solution().merge([[1, 8], [3, 6], [9, 11]]))

File: sort/56.merge-intervals.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
 public:
  vector<vector<int>> merge(vector<vector<int>>& intervals) {
    if (intervals.size() < 2) {
      return intervals;
    }

    sort(intervals.begin(), intervals.end());

    int n = intervals.size();
    vector<vector<int>> ans;
    for (const auto& interval : intervals) {
      int l = interval[0], r = interval[1];
      if (ans.empty() || ans.back()[1] < l) {
        ans.push_back({l, r});
      } else {
        ans.back()[1] = max(ans.back()[1], r);
      }
    }
    return ans;
  }
};

int main(int argc, char const* argv[]) {
  vector<vector<int>> intervals{{1, 8}, {3, 6}};
  auto ans = (new Solution)->merge(intervals);
  auto pv = [](const auto res) {
    for (auto&& i : res) cout << i << " ";
    cout << '\n';
  };
  for (auto&& i : ans) {
    pv(i);
  }

  return 0;
}
